app/controllers/cases/ico_foi_controller.rb

```diff
@@ -1,4 +1,111 @@
 module Cases
   class IcoFoiController < IcoController
+
+    before_action -> { set_decorated_case(params[:id]) }, only: [:record_further_action, :require_further_action, :confirm_record_further_action, :confirm_require_further_action]
+
+    def record_further_action
+      authorize @case, :can_record_further_action?
+    
+      set_permitted_events
+      read_info_from_session
+      @s3_direct_post = S3Uploader.for(@case, 'requests')
+      render 'cases/ico_foi/record_further_action'
+    end
+
+    def confirm_record_further_action
+      authorize @case, :can_record_further_action?
+      set_permitted_events
+      if params_valid?
+        session_persist_state(params_for_record_further_action)
+        redirect_to require_further_action_case_ico_foi_path(@case)
+      else
+        @case = @case.decorate
+        @s3_direct_post = S3Uploader.for(@case, 'requests')
+        render :record_further_action
+      end
+    end
+
+    def require_further_action
+      authorize @case, :can_require_further_action?
+    
+      set_permitted_events
+      render 'cases/ico_foi/require_further_action'
+    end
+
+    def confirm_require_further_action
+      authorize @case, :can_record_further_action?
+      set_permitted_events
+
+      service = CaseRequireFurtherActionService.new(
+        current_user,
+        @case,
+        params_for_requiring_further_action
+      )
+      result = service.call
+      if result == :ok
```

Comment:
Not a blocker but we could extract this case statement into a method as this method is already quite long and have something like:

```ruby
flash[:notice] = prepare_flash_message(case) if result == :ok
```